### PR TITLE
Add example event filter based on agent keeaplive timeout values

### LINKED ISSUE #2825: Add example alert fatigue event filter based on keepalive timeout flags
### Description
In #2800 we documented mapping between event attributes and the `keepalive-critical-timeout` and `keepalive-warning-timeout` flag and mentioned that these event attributes are useful for writing alert fatigue filters. Jef agrees that we should add an example of such a filter and is planning to develop one. When he does, we should add it to the docs.

### Category

This is an issue with:

- [ ] errata (error in doc content)
- [ ] new doc (requires a completely new doc)
- [ ] update (add missing or refresh existing content)
- [x] improvement (improve docs content or format)

### Affected pages
Example should probably be added to https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-filter/filters/#event-filter-examples

Add links to event filter example:
- https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-schedule/agent/#keepalive-monitoring
- https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-schedule/agent/#keepalive-configuration-flags

### Expected behavior
Docs explain that you can use keepalive-based event attributes to create event filters and provide an example

### Current behavior
Docs mention that you can use keepalive-based event attributes to create event filters but do not provide an example

### Corrections needed
Add keepalive-based event attribute event filter example in event filter reference and add links in agent reference to help readers find it

### PR DESCRIPTION
**Pull request by hillaryfraley**
## Description
- Adds an example event filter to reduce alert fatigue for keepalive events based on agent keepalive timeout values.
- Adds links to the example event filter in keepalive monitoring and keepalive flags sections of the agent reference.

## Motivation and Context
Closes https://github.com/sensu/sensu-docs/issues/2825

## Review Instructions
I'm not sure how well my attempt at the example filter matches what Jef had in mind for https://github.com/sensu/sensu-docs/issues/2825

### CODE REVIEW

**Comment by hillaryfraley** [comment]
Reading through https://bonsai.sensu.io/assets/nixwiz/sensu-go-fatigue-check-filter#keepalives and wondering if this example should use Todd's fatigue check filter asset instead.

**Comment by jspaleta** [comment]
We should probably make it clear the filter is intended to work with specific non-default values of keepalive timeout configuration.  This filter will only work if you set the warning timeout to 180 and the critical timeout to 300.   It will not work with the current (6.2.3) default configuration for agent that has warning set to 120 and critical disabled.

**Comment by jspaleta** [comment]
I'm actually not sure what the timeout line in the filtering is meant to achieve...
`event.check.timeout == 180 || event.check.timeout == 300`

So let me backup if the goal is to design a filter that will alert on first occurrence of the state then at most once every 15 minutes..I would build an expression like this:
```
 expressions:
   - event.check.occurrences == 1 ||  ( event.check.timeout > 0  && event.check.occurrences % parseInt( 60 * 15 / event.check.timeout ) == 0)
   - event.check.occurrences == 1 ||  ( event.check.ttl > 0  && event.check.occurrences % parseInt( 60 * 15 / event.check.ttl) == 0 )

```
The first expression works if the keepalive warning is configured (and by default this is set to 120 seconds)
The second expression works for the keepalive critical is configured (set to zero by default)

The parseInt() does the job of calculating the number of occurances that best match 15 minutes..truncating the calculation to an integer if needed.  So in the case of 120 warning timeout... its actually gonna be an alert every 14 minutes as best fit to 15 minute interval.

**Comment by jspaleta** [comment]
actually that is not correct either.. the critical keepalive timeout doesn't set the event rate.. that is only set by the warning timeout, because it maps to the check timeout

This is the sufficient logic I think based on my local tests.

```
expressions:
   - is_incident
   - event.check.occurrences == 1 || event.check.occurrences % parseInt( 60 * 15 / event.check.timeout ) == 0
```   

We can get really tricky i think and reference an entity attribute instead of that hardcoded 15 minutes

**Comment by hillaryfraley** [comment]
Thank you @jspaleta ! I was just trying to come up with an example that uses the event attributes that are based on the keepalive-critical-timeout and keepalive-warning-timeout.

**Comment by jspaleta** [comment]
Okay here's the ultimate version of this filter!

```
 expressions:
    - is_incident
    - "event.check.occurrences == 1 || event.check.occurrences % parseInt( 60 * ( 'keepalive_alert_minutes' in event.entity.annotations ? parseInt(event.entity.annotations.keepalive_alert_minutes): 15) / event.check.timeout ) == 0" 
```

Here I've added conditional javascript logic that will check for an entity level annotation "keepalive_alert_minutes" and if it exists will parse that as an integer. If it does not exist will fallback to using 15 minutes for the alert cadence.

File: content/sensu-go/6.1/observability-pipeline/observe-filter/filters.md [hunk 1]
```diff
@@ -943,6 +943,54 @@ spec:
...
+metadata:
+  annotations: null
+  labels: null
+  name: keepalive_timeouts
+  namespace: default
+spec:
+  action: allow
+  expressions:
+   - is_incident
+   - event.check.occurrences == 1 || ( event.check.timeout > 0  && event.check.status == 1 && event.check.occurrences % parseInt( 60 * 15 / event.check.timeout ) == 0)
```
Comment (hillaryfraley) [inline content/sensu-go/6.1/observability-pipeline/observe-filter/filters.md]:
```suggestion
   - event.check.occurrences == 1 || event.check.occurrences % parseInt( 60 * ( 'keepalive_alert_minutes' in event.entity.annotations ? parseInt(event.entity.annotations.keepalive_alert_minutes): 15) / event.check.timeout ) == 0
```

File: content/sensu-go/6.1/observability-pipeline/observe-filter/filters.md [hunk 2]
```diff
@@ -943,6 +943,54 @@ spec:
...
+    "name": "keepalive_timeouts",
+    "namespace": "default",
+    "labels": null,
+    "annotations": null
+  },
+  "spec": {
+    "action": "allow",
+    "expressions": [
+      "is_incident",
+      "event.check.occurrences == 1 || ( event.check.timeout > 0  && event.check.status == 1 && event.check.occurrences % parseInt( 60 * 15 / event.check.timeout ) == 0)"
```
Comment (hillaryfraley) [inline content/sensu-go/6.1/observability-pipeline/observe-filter/filters.md]:
```suggestion
      "event.check.occurrences == 1 || event.check.occurrences % parseInt( 60 * ( 'keepalive_alert_minutes' in event.entity.annotations ? parseInt(event.entity.annotations.keepalive_alert_minutes): 15) / event.check.timeout ) == 0"
```

File: content/sensu-go/6.1/observability-pipeline/observe-filter/filters.md [hunk 3]
```diff
@@ -943,6 +943,54 @@ spec:
 
 {{< /language-toggle >}}
 
+### Reduce alert fatigue for keepalive events
+
+This example `keepalive_timeouts` event filter will match event data with an occurrences value of 1 OR any occurrences value that matches 15 minutes via a modulo operator calculation.
+The `parseInt()` function calculates the number of occurences that match 15 minutes and truncates the calculation to an integer if needed.
+
+This event filter will limit keepalive timeout events to the first occurrence and every 15 minutes thereafter.
```
Comment (hillaryfraley) [inline content/sensu-go/6.1/observability-pipeline/observe-filter/filters.md]:
```suggestion
This example `keepalive_timeouts` event filter will match event data with an occurrences value of 1 OR any occurrences value that matches 15 minutes via a modulo operator calculation.
This limits keepalive timeout event alerts to the first occurrence and every 15 minutes thereafter.

This example uses conditional JavaScript logic to check for an entity-level annotation, `keepalive_alert_minutes`, and if it exists, parses the annotation value as an integer.
If the annotation does not exist, the event filter uses 15 minutes for the alert cadence. 
```

**Pull request merged**